[tech-onboarding assessment] Rapid File Sharing App with Gems

abhinvv1 opened the pull request:
[Problem Link](https://github.com/browserstack/tech-onboarding/tree/master/rapid-share-with-gems)
### Why?
- This PR is to implement an online file sharing application using Rails.

### What?
The Rapid sharing file has the following features:
- User Management:
   - New user will be able to register for a new account with a unique username, email address and password (greater than 8 characters, Include at least one uppercase letter, Include at least one lowercase letter and Include at least one letter), used `devise` Gem
   - Login and logout handling
- User Profile:
   - Profile page for every user where user can update various details about their account like: `name`, `email address`
- File Management:
   - File Dashboard: Contain various files associated with a user account organized in descending order of upload date
   - File Upload: A user is able to upload one file at a time. File is saved into `user_file` entity with file metadata such as `size` and `content-type`
   - File Download: A user can download file through download action.
   - File Share: A user can turn on public sharing of a file using a “ON/OFF” toggle control. Once public sharing is turned on, then a unique URL should be made available next to the toggle button which is publically accessible.
  - File Deletion: A user can delete the associated file.


### Test
- Rspec to write unit tests for models and controllers

### Usage
1. Clone the repo
2. Check you have ruby and rails installed
3. run `bundler install` to install gems
4. run `rails db:migrate` to execute migration
5. run `rails server` to start the server
6. To run unit tests run `rspec`

### Sample output
- Login page
<img width="1439" alt="Screenshot 2024-07-31 at 1 37 07 PM" src="https://github.com/user-attachments/assets/a1a379b2-8a66-4778-97a1-e387d6a2780d">

- Signup page
<img width="1440" alt="Screenshot 2024-07-31 at 1 37 20 PM" src="https://github.com/user-attachments/assets/e1f3412c-63b9-4279-bd95-ce3a56479ab1">


- Home page
<img width="1440" alt="Screenshot 2024-07-31 at 1 37 52 PM" src="https://github.com/user-attachments/assets/74925327-ced2-4c58-9196-8f9ee44b3205">


- Files Dashboard
<img width="1440" alt="Screenshot 2024-07-31 at 1 38 01 PM" src="https://github.com/user-attachments/assets/399f4888-d583-48c8-a7ff-20ca1b139f71">


- Upload new file page
<img width="1440" alt="Screenshot 2024-07-31 at 1 38 16 PM" src="https://github.com/user-attachments/assets/5e73e16e-2141-45f9-87d0-c316484f2653">

- Shared file page
<img width="1440" alt="Screenshot 2024-07-31 at 1 38 30 PM" src="https://github.com/user-attachments/assets/82574a67-fc7e-404d-a16a-1e30dd8a4f70">

WinstonPais commented:
bugs
- showing up as 6 chars in signup
- wrong banner on signup
- email regex fix

abhinvv1 commented:
bug fixes:
1. updated the password length to 8 characters
2. fixed wrong banner issue screenshot attached for reference
3. updated with stronger email regex 


<img width="1440" alt="Screenshot 2024-08-01 at 8 40 56 PM" src="https://github.com/user-attachments/assets/e6ba9f19-f375-4b76-a925-bac6c322e9f7">
